### PR TITLE
Reintroduce cover photo

### LINKED ISSUE #1823: Reintroduce cover photo
The possibility to upload a cover photo for APInf landing page has been removed from Branding settings. Since cover photo was part of APIKA landing page requirements, it should be reintroduced. Also, it gives more options for re-branding in general and also branding apinf.io.
# User story

> As an APInf administrator
> I want to use a cover photo on landing page
> to make my APInf deployment better match with other brand related web pages or tools.
# Definition of done
- [x] Admin can upload a cover photo from Branding settings.
- [x] If cover photo is uploaded, the landing page shows the cover photo in cover photo area.
- [x] Admin can replace a cover photo, if possible.
- [x] Admin can delete a cover photo.
  - [x] Cover photo area shows primary branding color after cover photo has been deleted.
  - [x] Image is removed from collection.

# Wireframes
apinfcoverphoto
apinfcoverphoto2

### PR DESCRIPTION
**Pull request by marla-singer**
Closes #1823

### CODE REVIEW

**Comment by jykae** [comment]
Reviewing

**Comment by jykae** [comment]
Basic functionality works nice 👍 

Using white background, title is hard to see, highlighted with mouse (request from @Nazarah, https://github.com/apinf/platform/issues/1823#issuecomment-257505098)
nayttokuva 2016-11-09 kello 16 13 46

**Comment by marla-singer** [comment]
I sugget to implement feature "Change site title color" in section Color. It wil provide to users to change color special for site title. But change it via automatic,well, I think it isn't good idea for UX

**Comment by jykae** [comment]
@marla-singer discussed with @bajiat site title color can be enhancement.

**Comment by jykae** [comment]
@marla-singer Can you check if there is some border left after deleting cover photo?

nayttokuva 2016-11-09 kello 16 37 11

**Comment by marla-singer** [comment]
@jykae Yes, I found some strange behaviour

**Comment by jykae** [comment]
@marla-singer what you think about adding preview of uploaded coverphoto like we have for site logo?

**Comment by marla-singer** [comment]
@jykae I thought about it. Why not? Let's do it :dancer:

**Comment by marla-singer** [comment]
@jykae Ready. Sorry, I wasn't attentive on wireframes

**Comment by jykae** [comment]
@marla-singer looks good, merging

File: branding/cover_photo/client/upload/upload.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { Template } from 'meteor/templating';
+import { Mongo } from 'meteor/mongo';
+import { ReactiveVar } from 'meteor/reactive-var';
+import { TAPi18n } from 'meteor/tap:i18n';
+import { sAlert } from 'meteor/juliancwirko:s-alert';
+// Import apinf collections
+import { Branding } from '/branding/collection';
+import CoverPhoto from '/branding/cover_photo/collection';
+
+const uploadingSpinner = new ReactiveVar(false);
```
Comment (brylie) [inline branding/cover_photo/client/upload/upload.js]: This is unconventional in our project. We usually assign reactive variables as Template.instance() variables. It seems a bit 'leaky' to rely on this reactive variable within a startup function.

File: branding/cover_photo/client/upload/resumable.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import { Branding } from '/branding/collection';
+import CoverPhoto from '/branding/cover_photo/collection';
+import { fileNameEndsWith } from '/core/helper_functions/file_name_ends_with';
+import uploadingSpinner from './upload';
+
+Meteor.startup(function () {
+  // Set cover photo id to branding collection on Success
+  CoverPhoto.resumable.on('fileSuccess', function (file) {
+    // Turn off spinner
+    uploadingSpinner.set(false);
```
Comment (brylie) [inline branding/cover_photo/client/upload/resumable.js]: Is there a different way to indicate that the file is uploading? This is relying on a reactive variable defined in a different file, so is a bit tricky to understand.

File: branding/cover_photo/client/view/view.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  // Subscribe to Branding collection
+  instance.subscribe('branding');
+  // Subscribe to Cover Photo collection
+  instance.subscribe('coverPhoto');
+});
+
+Template.viewCoverPhoto.helpers({
+  coverPhotoExists () {
+    // Get Branding collection
+    const branding = Branding.findOne();
```
Comment (brylie) [inline branding/cover_photo/client/view/view.js]: Branding.findOne() may return undefined in some cases.

File: branding/cover_photo/client/upload/upload.js (hunk 1 above)
Comment (marla-singer) [inline branding/cover_photo/client/upload/upload.js]: After that it was needed to export this variable and I couldn't think up another idea to do it

File: branding/cover_photo/client/upload/resumable.js (hunk 2 above)
Comment (marla-singer) [inline branding/cover_photo/client/upload/resumable.js]: Yes, it's really strange way but it was the better my idea how to indicate that the file is uploading

File: home/client/body/homeBody.html [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <template name="homeBody">
-  <div class="coverphoto">
+<div class="coverphoto" style="background-image: url({{coverPhotoUrl}})">
```
Comment (brylie) [inline home/client/body/homeBody.html]: This should probably be done in our dynamic stylesheet:

- https://github.com/apinf/platform/tree/develop/core/client/custom_stylesheet

File: home/client/body/homeBody.js [hunk 5]
```diff
@@ -29,4 +36,21 @@ Template.homeBody.helpers({
 
     return mailEnabled;
   },
+  coverPhotoUrl () {
+    // Get ID
+    const currentCoverPhotoFileId = this.branding.coverPhotoFileId;
```
Comment (brylie) [inline home/client/body/homeBody.js]: There may be cases where `this.branding` is undefined.

File: home/client/body/homeBody.js [hunk 6]
```diff
@@ -29,4 +36,21 @@ Template.homeBody.helpers({
...
+
+    // Convert to Mongo ObjectID
+    const objectId = new Mongo.Collection.ObjectID(currentCoverPhotoFileId);
```
Comment (brylie) [inline home/client/body/homeBody.js]: There may be cases in our code where `currentCoverPhotoFileId` is undefined. Would this potentially throw an exception?

File: branding/cover_photo/client/upload/resumable.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Get the id from project logo file object
+    const coverPhotoFileId = file.uniqueIdentifier;
+
+    // Get branding
+    const branding = Branding.findOne();
```
Comment (brylie) [inline branding/cover_photo/client/upload/resumable.js]: It is possible for this `Branding.findOne()` to return undefined, particularly when deploying a new project.

File: branding/cover_photo/client/upload/upload.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    return CoverPhoto.findOne(objectId);
+  },
+  spinnerEnabled () {
+    // Get value of spinner
+    return uploadingSpinner.get();
+  },
+});
+
+Template.uploadCoverPhoto.events({
+  'click .delete-cover-photo': function () {
```
Comment (brylie) [inline branding/cover_photo/client/upload/upload.js]: Great comments here. Good i18n keys. :smiley:

File: home/client/body/homeBody.js (hunk 6 above)
Comment (marla-singer) [inline home/client/body/homeBody.js]: if `currentCoverPhotoFileId` is undefined then `objectId` will be undefined too. So `currentCoverPhotoFile` will be also undefined and helper'll return empty line. I don't think that throw an exception needs

File: branding/cover_photo/client/upload/uploadButton/uploadButton.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+// Import meteor packages
+import { Template } from 'meteor/templating';
+// Import apinf collections
+import CoverPhoto from '/branding/cover_photo/collection';
+
+Template.uploadCoverPhotoButton.onRendered(function() {
+  // Assign resumable browse to element
+  CoverPhoto.resumable.assignBrowse($('#cover-photo-browse'));
```
Comment (brylie) [inline branding/cover_photo/client/upload/uploadButton/uploadButton.js]: We may be able to store a reference to this object, which might give us access to file events. That way, we don't need a global reactive variable for the Meteor.startup scope.

**Comment by brylie** [comment]
We need to revert this PR  from being merged. There are console errors when starting a fresh instance.

**Comment by marla-singer** [comment]
Sadly

**Comment by brylie** [comment]
@marla-singer can we open a quick hotfix for this branch?

**Comment by brylie** [comment]
E.e. we can leave the merge PR as it is, and just open a new pull request to fix the console errors. It would also be good to figure out a different approach to show/hide the spinner.

**Comment by brylie** [comment]
```js
Exception in template helper: TypeError: Cannot read property 'coverPhotoFileId' of undefined
    at Object.coverPhotoUrl (http://localhost:3000/app/app.js?hash=d5edf42d8ba275cd09fd48a0c5c5d5f52128f460:35425:50)
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3035:16
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:1697:16
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3087:66
    at Function.Template._withTemplateInstanceFunc (http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3728:12)
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3086:27
    at Spacebars.call (http://localhost:3000/packages/spacebars.js?hash=45ae773bea8216022dfca4f7eb46d878e5336593:172:18)
    at Spacebars.mustacheImpl (http://localhost:3000/packages/spacebars.js?hash=45ae773bea8216022dfca4f7eb46d878e5336593:106:25)
    at Object.Spacebars.mustache (http://localhost:3000/packages/spacebars.js?hash=45ae773bea8216022dfca4f7eb46d878e5336593:110:39)
    at style (http://localhost:3000/app/app.js?hash=d5edf42d8ba275cd09fd48a0c5c5d5f52128f460:35098:52)
```

File: home/client/body/homeBody.js (hunk 5 above)
Comment (brylie) [inline home/client/body/homeBody.js]:
```
Exception in template helper: TypeError: Cannot read property 'coverPhotoFileId' of undefined
    at Object.coverPhotoUrl (http://localhost:3000/app/app.js?hash=d5edf42d8ba275cd09fd48a0c5c5d5f52128f460:35425:50)
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3035:16
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:1697:16
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3087:66
    at Function.Template._withTemplateInstanceFunc (http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3728:12)
    at http://localhost:3000/packages/blaze.js?hash=a397812f1de3cd57d393a29b5e17e73a6d75de13:3086:27
    at Spacebars.call (http://localhost:3000/packages/spacebars.js?hash=45ae773bea8216022dfca4f7eb46d878e5336593:172:18)
    at Spacebars.mustacheImpl (http://localhost:3000/packages/spacebars.js?hash=45ae773bea8216022dfca4f7eb46d878e5336593:106:25)
    at Object.Spacebars.mustache (http://localhost:3000/packages/spacebars.js?hash=45ae773bea8216022dfca4f7eb46d878e5336593:110:39)
    at style (http://localhost:3000/app/app.js?hash=d5edf42d8ba275cd09fd48a0c5c5d5f52128f460:35098:52)
```

File: branding/cover_photo/client/upload/resumable.js (hunk 7 above)
Comment (marla-singer) [inline branding/cover_photo/client/upload/resumable.js]: Nope, in this fragment we are sure that branding collections exists https://github.com/apinf/platform/blob/e4429acc7206dbbb32e61acf1b7ad72372c26349/branding/client/branding.html#L14-L18

**Pull request merged**